src/app/ReadClient.cpp

```diff
@@ -134,85 +126,68 @@ CHIP_ERROR ReadClient::SendReadRequest(ReadPrepareParams & aReadPrepareParams)
         }
 
         if (aReadPrepareParams.mEventPathParamsListSize != 0 && aReadPrepareParams.mpEventPathParamsList != nullptr)
         {
             EventPathIBs::Builder & eventPathListBuilder = request.CreateEventRequests();
-            SuccessOrExit(err = request.GetError());
-            err = GenerateEventPaths(eventPathListBuilder, aReadPrepareParams.mpEventPathParamsList,
-                                     aReadPrepareParams.mEventPathParamsListSize);
-            SuccessOrExit(err);
+            ReturnErrorOnFailure(err = request.GetError());
+
+            ReturnErrorOnFailure(GenerateEventPaths(eventPathListBuilder, aReadPrepareParams.mpEventPathParamsList,
+                                                    aReadPrepareParams.mEventPathParamsListSize));
+
             if (aReadPrepareParams.mEventNumber != 0)
             {
                 // EventFilter is optional
                 EventFilterIBs::Builder & eventFilters = request.CreateEventFilters();
-                SuccessOrExit(err = request.GetError());
+                ReturnErrorOnFailure(request.GetError());
+
                 EventFilterIB::Builder & eventFilter = eventFilters.CreateEventFilter();
-                SuccessOrExit(err = eventFilters.GetError());
-                eventFilter.EventMin(aReadPrepareParams.mEventNumber).EndOfEventFilterIB();
-                SuccessOrExit(err = eventFilter.GetError());
-                eventFilters.EndOfEventFilters();
-                SuccessOrExit(err = eventFilters.GetError());
+                ReturnErrorOnFailure(eventFilters.GetError());
+                ReturnErrorOnFailure(eventFilter.EventMin(aReadPrepareParams.mEventNumber).EndOfEventFilterIB().GetError());
+                ReturnErrorOnFailure(eventFilters.EndOfEventFilters().GetError());
             }
         }
 
-        request.IsFabricFiltered(false).EndOfReadRequestMessage();
-        SuccessOrExit(err = request.GetError());
-
-        err = writer.Finalize(&msgBuf);
-        SuccessOrExit(err);
+        ReturnErrorOnFailure(request.IsFabricFiltered(false).EndOfReadRequestMessage().GetError());
+        ReturnErrorOnFailure(writer.Finalize(&msgBuf));
     }
 
     mpExchangeCtx = mpExchangeMgr->NewContext(aReadPrepareParams.mSessionHandle, this);
-    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
     mpExchangeCtx->SetResponseTimeout(aReadPrepareParams.mTimeout);
 
-    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReadRequest, std::move(msgBuf),
-                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReadRequest, std::move(msgBuf),
+                                                    Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)));
```

Comment:
If this fails, we need to close the exchange, no?  Or is the assumption that our consumer will destroy us and we will close it via that mechanism?

Reply:
The latter. Until ReadRequest completes successfully, the ReadClient doesn't take on shared responsibility for managing the closure of the exchange context on completion of work.